### PR TITLE
saltpack specs for Version 2

### PR DESCRIPTION
**Pull request by oconnor663**
r? @mlsteele @maxtaco

### CODE REVIEW

**Comment by oconnor663** [comment]
The easiest way to render these is to clone https://github.com/keybase/saltpack.org, point `views/docs_symlink` to this repo, and run `npm i && iced app.iced` to start saltpack.org locally.

File: specs/saltpack_encryption_v2.md [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+The main building block of our encrypted message format is NaCl's
+[box](http://nacl.cr.yp.to/box.html) and
+[secretbox](http://nacl.cr.yp.to/secretbox.html) constructions. These have
+several properties that we'll want to keep:
+- Privacy. If Alice sends a message to Bob, and Mallory (the man-in-the-middle)
+  intercepts the message, Mallory can't read any of its contents.
+- Authenticity. When Bob receives a message from Alice, he can verify that
+  Alice sent it, and that Mallory didn't change its contents.
+- Repudiability. It's possible for Bob to forge a message that appears to be
```
Comment (maxtaco) [inline specs/saltpack_encryption_v2.md]: still true?

File: specs/saltpack_encryption_v2.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+each message needs an ephemeral sender public key, used only for this one message,
+to hide the sender's true identity. Some implementations of the scheme can
+choose to reveal the keys of the receivers to make user-friendlier error
+messages on decryption failures (e.g., "Can't decrypt this message on your
+phone, but try your laptop.")  If the sender wants to decrypt the message
+at a later date, she simply adds her public keys to the list of recipients.
+
+## Implementation
+
+An encrypted message is a series of concatenated MessagePack objects. The first
```
Comment (maxtaco) [inline specs/saltpack_encryption_v2.md]: link for msgpack?

File: specs/saltpack_signcryption_v2.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    ephemeral public key,
+    sender secretbox,
+    recipients list,
+]
+```
+
+- The **format name** is the string "saltpack".
+- The **version** is a list of the major and minor versions, currently
+  `[2, 0]`. Note that saltpack version 1 did not include a signcryption mode.
+- The **mode** is the number 3, for encryption. (1 and 2 are attached and
```
Comment (maxtaco) [inline specs/saltpack_signcryption_v2.md]: check this, there's a bug somewhere in this statement!

File: specs/saltpack_signcryption_v2.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+[
+    recipient identifier,
+    payload key box,
+]
+```
+
+- The **recipient identifier** depends on the type of recipient key. For an
+  asymmetric recipient key, the identifier is derived from shared DH output.
+  For a symmetric key, the identifier is left to the application.
+- The **payload key box** is a [`crypto_box`](http://nacl.cr.yp.to/box.html)
```
Comment (maxtaco) [inline specs/saltpack_signcryption_v2.md]: what happens here in the case of symmetric key?

File: specs/saltpack_signcryption_v2.md (hunk 3 above)
Comment (maxtaco) [inline specs/saltpack_signcryption_v2.md]: I think `3` is for **sign**cryption?

File: specs/saltpack_signcryption_v2.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+```yaml
+# header packet (on the wire, this is twice-encoded)
+[
+  # format name
+  "saltpack",
+  # major and minor version
+  [1, 0],
+  # mode (0 = encryption)
+  0,
```
Comment (maxtaco) [inline specs/saltpack_signcryption_v2.md]: not `3` for **sign**cryption?

File: specs/saltpack_encryption_v2.md (hunk 1 above)
Comment (oconnor663) [inline specs/saltpack_encryption_v2.md]: Still true of the encryption format, not true of the signcryption format.

File: specs/saltpack_signcryption_v2.md (hunk 3 above)
Comment (oconnor663) [inline specs/saltpack_signcryption_v2.md]: Yep, forgot to change the word "encryption".

File: specs/saltpack_signcryption_v2.md (hunk 4 above)
Comment (oconnor663) [inline specs/saltpack_signcryption_v2.md]: This is where we hash together a bunch of metadata and some random bytes and post it to the server, where it winds up in the `tlf_pseudonyms` table and gets resolved by readers with the server's help. I didn't feel like that kind of stuff belonged in the saltpack spec. Do you think we should include it?

File: specs/saltpack_signcryption_v2.md (hunk 5 above)
Comment (oconnor663) [inline specs/saltpack_signcryption_v2.md]: yep, forgot to update

File: specs/saltpack_signcryption_v2.md (hunk 4 above)
Comment (oconnor663) [inline specs/saltpack_signcryption_v2.md]: Going to land as is, but let me know what you think about these details, and I can make a followup PR.

**Pull request merged**